Next.js/Gatsby Migration guides to Astro (#2211)
withastro/docs: src/content/docs/en/guides/migrate-to-astro/from-gatsby.mdx

title: Migrating from Gatsby
description: Tips for migrating an existing Gatsby project to Astro
type: migration
stub: false
framework: Gatsby
---
import AstroJSXTabs from '~/components/tabs/AstroJSXTabs.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

Here are some key concepts and migration strategies to help you get started. Use the rest of our docs and our [Discord community](https://astro.build/chat) to keep going!

## Key Similarities between Gatsby and Astro

Gatsby and Astro share some similarities that will help you migrate your project:

> - The [syntax of `.astro` files is similar to JSX](/en/core-concepts/astro-components/#differences-between-astro-and-jsx). Writing Astro should feel familiar.

Reviewer: ```suggestion
- The [syntax of `.astro` files is similar to JSX](/en/core-concepts/astro-syntax/#jsx-like-expressions). Writing Astro should feel familiar.
```

You can pass a `--template` argument to the `create astro` command to start a new Astro project with one of our official starters (e.g. `docs`, `blog`, `portfolio`). Or, you can [start a new project from any existing Astro repository on GitHub](/en/install/auto/#starter-templates).

  <PackageManagerTabs>
    <Fragment slot="npm">
    ```shell
    # launch the Astro CLI Wizard 
    npm create astro@latest

    # create a new project with an official example
    npm create astro@latest -- --template <example-name>
    ```
    </Fragment>
    <Fragment slot="pnpm">
    ```shell
    # launch the Astro CLI Wizard 
    pnpm create astro@latest

    # create a new project with an official example
    pnpm create astro@latest --template <example-name>
    ```
    </Fragment>
    <Fragment slot="yarn">
    ```shell
    # launch the Astro CLI Wizard 
    yarn create astro@latest

    # create a new project with an official example
    yarn create astro@latest --template <example-name>
    ```
    </Fragment>
  </PackageManagerTabs>

Then, copy your existing Gatsby project files over to your new Astro project into a separate folder outside of `src`.

:::tip

> Visit https://astro.new for the full list of official starter templates, and links for opening a new project in StackBlitz, CodeSandbox or Gitpod.

Reviewer 2: Missing serial comma
```suggestion
Visit https://astro.new for the full list of official starter templates, and links for opening a new project in StackBlitz, CodeSandbox, or Gitpod.
```

3. **Copy or Move** Gatsby's other files and folders (e.g. `components`, `pages`, etc.) as needed into your Astro `src/` folder as you rebuild your site, following [Astro's project structure](/en/core-concepts/project-structure/).

  Astro's `src/pages/` folder is a special folder used for file-based routing to create your site's pages and posts from `.astro`, `.md` and `.mdx` files. You will not have to configure any routing behavior for your Astro, Markdown, and MDX files.

  All other folders are optional, and you can organize the contents of your `src/` folder any way you like. Other common folders in Astro projects include `src/layouts/`, `src/components`, `src/styles`, and `src/scripts`.

### Tips: Convert JSX files to `.astro` files

Here are some tips for converting a Gatsby `.js` component into a `.astro` component:

1. Use only the `return()` of the existing Gatsby component function as your HTML template.

1. Change any [Gatsby or JSX syntax to Astro syntax](#reference-convert-to-astro-syntax) or to HTML web standards. This includes `<Link to="">`, `{children}`, and `className`, for example.

> 1. Move any necessary JavaScript, including import statements, into a ["code fence" (`---`)](/en/core-concepts/astro-components/#the-component-script). Note: JavaScript to [conditionally render content](/en/core-concepts/astro-components/#dynamic-html) is often written inside the HTML template directly in Astro.

Reviewer: ```suggestion
1. Move any necessary JavaScript, including import statements, into a ["code fence" (`---`)](/en/core-concepts/astro-components/#the-component-script). Note: JavaScript to [conditionally render content](/en/core-concepts/astro-syntax/#dynamic-html) is often written inside the HTML template directly in Astro.
```

See Astro's [testing guide](/en/guides/testing/) for more.

### Repurpose config files

Gatsby has several top-level configuration files that also include site and page metadata and are used for routing. You will not use any of these `gatsby-*.js` files in your Astro project, but there may be some content that you can reuse as you build your Astro project:

- `gatsby-config.js`: Move your `siteMetadata: {}` into `src/data/siteMetadata.js` (or `siteMetadata.json`) to import data about your site (title, description, social accounts, etc.) into page layouts.

- `gatsby-browser.js`: Consider adding anything used here directly into your [main layout](#migrating-layout-files)'s `<head>` tag.

- `gatsby-node.js`: You will not need to create your own nodes in Astro, but viewing the schema in this file may help you with defining types in your Astro project.

- `gatsby-ssr.js`: If you choose to use SSR in Astro, you will [add and configure the SSR adapter](/en/guides/server-side-rendering/#adding-an-adapter) of your choice directly in `astro.config.mjs`.

## Reference: Convert to Astro Syntax

> The following are some examples of Gatsby-specific syntax that you will need to convert to Astro. See more [differences between Astro and JSX](/en/core-concepts/astro-components/#differences-between-astro-and-jsx) in the guide to writing Astro components.

Reviewer: ```suggestion
The following are some examples of Gatsby-specific syntax that you will need to convert to Astro. See more [differences between Astro and JSX](/en/core-concepts/astro-syntax/#differences-between-astro-and-jsx) in the guide to writing Astro components.
```

```astro title="src/components/Card.astro" del={1} ins={2}
<div style={{backgroundColor: `#f4f4f4`, padding: `1em`}}>{message}</div>
<div style="background-color: #f4f4f4; padding: 1em;">{message}</div>
```

Tailwind is supported after installing the [Tailwind integration](/en/guides/integrations-guide/tailwind/). No changes to your existing Tailwind code are required!

Global styling is achieved in Gatsby using CSS imports in `gatsby-browser.js`. In Astro, you will import `.css` files directly into a main layout component to achieve global styles.

See more about [Styling in Astro](/en/guides/styling/).

### Gatsby Image Plugin to Astro

Convert Gatsby's `<StaticImage />` and `<GatsbyImage />` components with [Astro's own image integration components](/en/guides/images/#astros-image-integration), or with a standard HTML `<img>` / JSX `<img />` tag as appropriate.

```astro title="src/pages/index.astro"
---
import { Image } from '@astrojs/image/components';
import localImage from "../images/brand/logo.png";
import rocket from '../images/rocket.svg';
---
<Image src={localImage} alt="The Astro Logo" />
<img src={rocket} alt="A rocketship in space.">
```

> Astro's `<Image />` and `<Picture />` components are experimental, and currently work in `.astro` and `.mdx` files only. See a [full list of component attributes](/en/guides/integrations-guide/image/#usage) available for these components, and note that several will differ from Gatsby's attributes.

Reviewer 2: Unnecessary comma
```suggestion
Astro's `<Image />` and `<Picture />` components are experimental and currently work in `.astro` and `.mdx` files only. See a [full list of component attributes](/en/guides/integrations-guide/image/#usage) available for these components, and note that several will differ from Gatsby's attributes.
```
